cpp/src/copying/scatter.cu

```diff
@@ -0,0 +1,374 @@
+
+  // Compute block size
+  int grid_size, block_size;
+  auto bitmask_kernel = gather_bitmask_kernel<true, decltype(gather_map)>;
+  CUDA_TRY(cudaOccupancyMaxPotentialBlockSize(&grid_size, &block_size, bitmask_kernel));
+
+  auto valid_counts = thrust::device_vector<size_type>(target.size());
+  bitmask_kernel<<<grid_size, block_size, 0, stream>>>(*device_source,
+    gather_map, *device_target, valid_counts.data().get());
+
+  // TODO for_each with a zip iterator?
+  auto valid_counts_host = thrust::host_vector<size_type>(valid_counts);
+  for (size_t i = 0; i < target.size(); ++i) {
+    auto const& target_col = target[i];
+    if (target_col->nullable()) {
+      target_col->set_null_count(target_col->size() - valid_counts_host[i]);
+    }
+  }
+}
+
+template <typename index_type>
+struct column_scatterer {
+  template <typename T, std::enable_if_t<is_fixed_width<T>()>* = nullptr>
+  std::unique_ptr<column> operator()(column_view const& source,
+      column_view const& scatter_map, column_view const& target,
+      rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+  {
+    // Result column must have a null mask if the source does
+    auto result = copy_with_nullable(target, source.nullable(), mr, stream);
+    auto result_view = result->mutable_view();
+
+    // Transform negative indices to index + target size
+    auto scatter_iter = thrust::make_transform_iterator(
+      scatter_map.begin<index_type>(),
+      index_converter<index_type>{target.size()});
+
+    // NOTE use source.begin + scatter_map.size rather than end in case the
+    // scatter map is smaller than the number of source rows
+    thrust::scatter(rmm::exec_policy(stream)->on(stream), source.begin<T>(),
+      source.begin<T>() + scatter_map.size(), scatter_iter,
+      result_view.begin<T>());
+
+    return result;
+  }
+
+  template <typename T, std::enable_if_t<not is_fixed_width<T>()>* = nullptr>
+  std::unique_ptr<column> operator()(column_view const& source,
+      column_view const& scatter_map, column_view const& target,
+      rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+  {
+    CUDF_FAIL("Scatter column type must be fixed width");
+  }
+};
+
+struct scatter_impl {
+  template <typename T, std::enable_if_t<std::is_integral<T>::value
+     and not std::is_same<T, bool8>::value>* = nullptr>
+  std::unique_ptr<table> operator()(
+      table_view const& source, column_view const& scatter_map,
+      table_view const& target, bool check_bounds,
+      rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+  {
+    if (check_bounds) {
+      auto const begin = -target.num_rows();
+      auto const end = target.num_rows();
+      auto bounds = bounds_checker<T>{begin, end};
+      CUDF_EXPECTS(thrust::all_of(rmm::exec_policy(stream)->on(stream),
+        scatter_map.begin<T>(), scatter_map.end<T>(), bounds),
+        "Scatter map index out of bounds");
+    }
+
+    // TODO create separate streams for each col and then sync with master?
```

Comment:
Not until we have a stream pool, please.

Reply:
This was more of a reminder for future improvement. I'll just remove it.